onnxruntime/contrib_ops/cpu/transformers/subgraph_whisper_decoder.cc

```diff
@@ -0,0 +1,274 @@
+    ORT_RETURN_IF(num_subgraph_inputs < 7 + first_past_input_index_ ||
+                      (num_subgraph_inputs - first_past_input_index_ - 3) % 4 != 0,
+                  "number of inputs expected to be kFirstPastInputIndex + 4 * layers + 3, got:", num_subgraph_inputs);
+  } else {
+    ORT_RETURN_IF(num_subgraph_inputs < 5 + first_past_input_index_ ||
+                      (num_subgraph_inputs - first_past_input_index_ - 1) % 4 != 0,
+                  "number of inputs expected to be kFirstPastInputIndex + 4 * layers + 1, got:", num_subgraph_inputs);
+  }
+
+  ORT_RETURN_IF(num_subgraph_outputs < 3 || (num_subgraph_outputs - first_present_output_index_) % 2 != 0,
+                "number of outputs expected to be 1 + 2 * layers, got:", num_subgraph_outputs);
+
+  ORT_RETURN_IF(subgraph_inputs[0]->Name() != "input_ids",
+                "decoder subgraph input 0 shall be named as input_ids, got: ", subgraph_inputs[0]->Name());
+  if (first_past_input_index_ == 2) {
+    ORT_RETURN_IF(subgraph_inputs[1]->Name() != "encoder_hidden_states",
+                  "decoder subgraph input 1 shall be named as encoder_hidden_states, got: ",
+                  subgraph_inputs[1]->Name());
+  }
+
+  // check subgraph outputs
+  ORT_RETURN_IF(subgraph_outputs[0]->Name() != "logits",
+                "decoder subgraph output 0 shall be named as logits, got: ", subgraph_outputs[0]->Name());
+
+  const ONNX_NAMESPACE::TensorShapeProto* logits_shape = subgraph_outputs[0]->Shape();
+  const ONNX_NAMESPACE::TensorShapeProto* past_shape = subgraph_outputs[first_present_output_index_]->Shape();
+
+  // Save parameters related to the subgraph.
+  ORT_RETURN_IF_ERROR(GetParameters(past_shape, logits_shape, false));
+  num_layers = (static_cast<int>(subgraph_outputs.size()) - first_present_output_index_) / 2;
+
+  // If input_ids's shape is ['batch_size', 1] then use next token as input_ids.
+  // Otherwise in the case of shape ['batch_size', 'sequence'], use sequence as input_ids.
+  const ONNX_NAMESPACE::TensorShapeProto* input_ids_shape = subgraph_inputs[0]->Shape();
+  if (input_ids_shape->dim(1).has_dim_value() && input_ids_shape->dim(1).dim_value() == 1) {
+    use_sequence_as_input_ids_ = false;
+  }
+
+  constexpr auto int32_type = ONNX_NAMESPACE::TensorProto_DataType::TensorProto_DataType_INT32;
+  constexpr auto float32_type = ONNX_NAMESPACE::TensorProto_DataType::TensorProto_DataType_FLOAT;
+  constexpr auto float16_type = ONNX_NAMESPACE::TensorProto_DataType::TensorProto_DataType_FLOAT16;
+
+  ORT_RETURN_IF(subgraph_inputs[0]->TypeAsProto()->tensor_type().elem_type() != int32_type,
+                "decoder subgraph input 0 (input_ids) shall have int32 type");
+
+  auto float_type = subgraph_inputs[2]->TypeAsProto()->tensor_type().elem_type();
+  ORT_RETURN_IF(float_type != float32_type && float_type != float16_type,
+                "decoder subgraph input 2 (encoder_hidden_states) shall have float or float16 type");
```

Comment:
>input 2 (encoder_hidden_states) [](http://example.com/codeflow?start=34&length=31)

encoder_hidden_states doesn't exist or will be input 1.

Reply:
Will fix in another PR